### PR TITLE
Siteless checkout: Allow to connect site after checkout.

### PR DESCRIPTION
**Pull request by elliottprogrammer**
This PR adds the ability to connect the user's Jetpack site after product purchase (checkout), if given the appropriate query args in the siteless (license-based) checkout url.

This PR can be tested along with Jetpack PR https://github.com/Automattic/jetpack/pull/33257 .

## Proposed Changes

* Updated the checkout flow to receive 3 new query args (`connect_after_checkout`, `admin_url`, and `from_site_slug`) - My Jetpack, in the plugin, will be passing these query args into checkout.
* When these query args are received in the siteless chekout flow, checkout will redirect to the site's connection/authorization url after purchase is completed.

Still TODO: Then after connection complete, activate the product license key (transfer the temporary site subscription to the target site) - In a separate, follow-up diff/PR.

## Testing Instructions

* Follow testing instructions in Jetpack PR https://github.com/Automattic/jetpack/pull/33257

#### Unit tests:

Run:

`yarn test-client client/my-sites/checkout/get-thank-you-page-url/test/get-thank-you-page-url.ts`
`yarn test-client client/my-sites/checkout/src/test/pending-page.ts`

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/my-sites/checkout/checkout-thank-you/pending/index.tsx [hunk 1]
```diff
@@ -37,6 +37,7 @@ interface CheckoutPendingProps {
 	receiptId: number | undefined;
 	siteSlug?: string;
 	redirectTo?: string;
+	fromSite?: string;
```
Comment (sirbrillig) [inline client/my-sites/checkout/checkout-thank-you/pending/index.tsx]: Could you add a comment above this definition (and all the other places `fromSite`/`fromSiteSlug` is used) to explain what it means? I think that "fromSite" is ambiguous enough that it would benefit from having an explanation so it doesn't get misused in the future.

```suggestion
    /**
    * I don't know what this means...
    */
	fromSite?: string;
```

File: client/my-sites/checkout/get-thank-you-page-url/index.ts [hunk 2]
```diff
@@ -228,6 +232,13 @@ export default function getThankYouPageUrl( {
 		}
 
 		// siteless checkout
+		if ( connectAfterCheckout && adminUrl && fromSiteSlug ) {
+			debug( 'Redirecting to the site to initiate Jetpack connection' );
+			// TODO: Then after connection, transfer temporary site subscription to the target site.
+			const connectUrl = `${ adminUrl }/admin.php?page=jetpack&connect_url_redirect&from=my-jetpack&redirect_after_auth=${ adminUrl }/admin.php?page=my-jetpack#/add-license`;
```
Comment (sirbrillig) [inline client/my-sites/checkout/get-thank-you-page-url/index.ts]: All the outputs of this function are covered by tests (see the top of this file for details). Could you add a test for this case?

File: client/my-sites/checkout/src/lib/pending-page.ts [hunk 3]
```diff
@@ -335,7 +345,7 @@ export function getRedirectFromPendingPage( {
 					redirectTo ?? getDefaultSuccessUrl( siteSlug, receiptId ),
 					receiptId
 				),
-				siteSlug,
+				siteSlug || fromSite,
```
Comment (sirbrillig) [inline client/my-sites/checkout/src/lib/pending-page.ts]: I don't fully understand the use of `fromSite` but should this be added to the other return values of this function as well? Most purchases don't have a receipt ID at this stage and will end up in the `if ( transaction )` block below.

File: client/my-sites/checkout/src/lib/pending-page.ts [hunk 4]
```diff
@@ -170,6 +178,7 @@ export function addUrlToPendingPageRedirect(
 	const successUrlObject = new URL( successUrlBase );
 	successUrlObject.searchParams.set( 'redirectTo', url );
 	successUrlObject.searchParams.set( 'receiptId', String( receiptId ) );
+	fromSiteSlug && successUrlObject.searchParams.set( 'from_site', fromSiteSlug );
```
Comment (sirbrillig) [inline client/my-sites/checkout/src/lib/pending-page.ts]: Using the `&&` pattern is fine inside expressions where it is necessary to be short, but I think here the code would be more readable with a full `if` statement.

File: client/my-sites/checkout/src/lib/pending-page.ts [hunk 5]
```diff
@@ -23,6 +24,7 @@ export interface RedirectForTransactionStatusArgs {
 	redirectTo?: string;
 	siteSlug?: string;
 	saasRedirectUrl?: string;
+	fromSite?: string;
```
Comment (sirbrillig) [inline client/my-sites/checkout/src/lib/pending-page.ts]: What's the difference between `fromSite` and `fromSiteSlug`? 🤔

**Review by CodeyGuyDylan — CHANGES_REQUESTED** [review]
I think these changes look good and they tested well with the plugin PR, but I will Request Changes for the same reason as the other one in that I have a question about potentially automatically assigning the license to the site since we do have the blog URL that they came from when trying to purchase

**Comment by CodeyGuyDylan** [comment]
Disregard my previous comment, I see now that the product activation will be added later on. I will leave in Requested Changes for now so you can address @sirbrillig's comments

File: client/my-sites/checkout/checkout-thank-you/pending/index.tsx [hunk 6]
```diff
@@ -72,6 +73,7 @@ function CheckoutPending( {
 	receiptId,
 	siteSlug,
 	redirectTo,
+	fromSite,
```
Comment (jboland88) [inline client/my-sites/checkout/checkout-thank-you/pending/index.tsx]: would it make sense to use the same parameter name for `fromSite` and `fromSiteSlug`? (If they are the same thing that is 🤔 )

File: client/my-sites/checkout/controller.jsx [hunk 7]
```diff
@@ -46,7 +46,15 @@ import { getProductSlugFromContext, isContextJetpackSitelessCheckout } from './u
...
 export function checkoutJetpackSiteless( context, next ) {
-	sitelessCheckout( context, next, { sitelessCheckoutType: 'jetpack' } );
+	const connectAfterCheckout = context.query?.connect_after_checkout === 'true';
+	const fromSiteSlug = context.query?.from_site_slug;
+	const adminUrl = context.query?.admin_url;
+	sitelessCheckout( context, next, {
+		sitelessCheckoutType: 'jetpack',
+		connectAfterCheckout,
+		...( fromSiteSlug && { fromSiteSlug } ),
+		...( adminUrl && { adminUrl } ),
```
Comment (jboland88) [inline client/my-sites/checkout/controller.jsx:65]: Curious - what's the advantage of using a spread operator with a conditional here? Would it work to just pass `fromSiteSlug` and `adminUrl` directly?

File: client/my-sites/checkout/src/hooks/use-create-payment-complete-callback.tsx [hunk 8]
```diff
@@ -64,6 +64,9 @@ export default function useCreatePaymentCompleteCallback( {
 	siteSlug,
 	sitelessCheckoutType,
 	checkoutFlow,
+	connectAfterCheckout,
+	adminUrl: wpAdminUrl,
```
Comment (jboland88) [inline client/my-sites/checkout/src/hooks/use-create-payment-complete-callback.tsx:68]: does wpAdminUrl need to be pulled from the state for this component?

File: client/my-sites/checkout/src/hooks/use-create-payment-complete-callback.tsx [hunk 9]
```diff
@@ -76,13 +79,16 @@ export default function useCreatePaymentCompleteCallback( {
...
+	adminUrl?: string;
+	fromSiteSlug?: string;
 } ): PaymentEventCallback {
 	const cartKey = useCartKey();
 	const { responseCart, reloadFromServer: reloadCart } = useShoppingCart( cartKey );
 	const reduxDispatch = useDispatch();
 	const siteId = useSelector( getSelectedSiteId );
 	const selectedSiteData = useSelector( getSelectedSite );
-	const adminUrl = selectedSiteData?.options?.admin_url;
+	const adminUrl = selectedSiteData?.options?.admin_url || wpAdminUrl;
```
Comment (jboland88) [inline client/my-sites/checkout/src/hooks/use-create-payment-complete-callback.tsx:98]: If `adminUrl` is re-defined here, does it need to be passed in the props?

File: client/my-sites/checkout/src/hooks/use-get-thank-you-url/index.tsx [hunk 10]
```diff
@@ -36,10 +36,13 @@ export default function useGetThankYouUrl( {
...
 	isInModal,
 	domains,
+	connectAfterCheckout,
+	adminUrl: wpAdminUrl,
+	fromSiteSlug,
 }: GetThankYouUrlProps ): GetThankYouUrl {
 	const selectedSiteData = useSelector( ( state ) => getSelectedSite( state ) );
 
-	const adminUrl = selectedSiteData?.options?.admin_url;
+	const adminUrl = selectedSiteData?.options?.admin_url || wpAdminUrl;
```
Comment (jboland88) [inline client/my-sites/checkout/src/hooks/use-get-thank-you-url/index.tsx:45]: If `adminUrl` is defined here, does it need to be passed in props?
Also, does wpAdminUrl need to be collected from the state? I don't see that variable defined here 🤔

**Review by jboland88 — COMMENTED** [review]
This tested well - I just noted a few small things inline.

I did notice that I was not redirected to the add-license page after authorizing connection, it looks like that was the intended final destination 🤔 That's not a blocker though IMO.

File: client/my-sites/checkout/src/lib/pending-page.ts (hunk 5 above)
Comment (elliottprogrammer) [reply client/my-sites/checkout/src/lib/pending-page.ts]: > What's the difference between `fromSite` and `fromSiteSlug`? 🤔

`fromSiteSlug` and `fromSite` are both the same value- the site's slug(ultimately originating from the 'not-connected yet' plugin). The only difference is that `fromSiteSlug` comes from a url query arg on route `/checkout/jetpack/:productSlug` which then get's passed along as a url query arg `fromSite` on route `/checkout/thank-you/no-site/pending/:orderId`.
I think I named them slightly different so as to distinguish that they are defined from url query args from different routes.., but maybe this naming difference is not necessary and actually more confusing.
I'll rename `fromSite` to `fromSiteSlug` keeping the name consistent. 👍

File: client/my-sites/checkout/src/hooks/use-get-thank-you-url/index.tsx (hunk 10 above)
Comment (elliottprogrammer) [reply client/my-sites/checkout/src/hooks/use-get-thank-you-url/index.tsx:45]: > If adminUrl is defined here, does it need to be passed in props?

No, In the "connectAfterCheckout" flow, the site will not be in context (because it has not been connected to wordpress.com yet), and therefore attempting to retrieve the site's adminUrl from state will return `undefined` (again because there will be no site in context), so to remedy this we are getting the adminUrl from a checkout url query arg and passing it in as a prop here.
In other words, in the 'connectAfterCheckout' flow(site is not in context), the site's wp-admin url is passed in here as a prop, but in any other checkout flows where the site is in context, the site's wp-admin url is retrieved from state. 🙂

> Also, does wpAdminUrl need to be collected from the state? I don't see that variable defined here

No, in the 'connectAfterCheckout' flow the site's wp-admin url is passed into this component as prop `adminUrl` but renamed to `wpAdminUrl`. Then adminUrl is defined loosely as `const adminUrl = state.admin_url || props.wpAdminUrl`, if that makes sense..

File: client/my-sites/checkout/get-thank-you-page-url/index.ts (hunk 2 above)
Comment (elliottprogrammer) [reply client/my-sites/checkout/get-thank-you-page-url/index.ts]: Yes of course, I should know better, my bad.  I have updated/modified this file many times and I know I need to add test coverage, and I was meaning to..., but then got to working on things and it slipped my mind. My apologies.  Yes, test coverage coming right up! 👍 😉

File: client/my-sites/checkout/controller.jsx (hunk 7 above)
Comment (elliottprogrammer) [reply client/my-sites/checkout/controller.jsx:65]: > Curious - what's the advantage of using a spread operator with a conditional here? Would it work to just pass `fromSiteSlug` and `adminUrl` directly?

Yes, I guess no advantage really..  My thinking was, if the query args are present in the url, then pass the props, otherwise don't.  I guess the only difference would be that if you were  inspecting `extraProps` at this point, you would only see properties that are defined (with values), there wouldn't be properties present with value 'undefined'.

File: client/my-sites/checkout/checkout-thank-you/pending/index.tsx (hunk 1 above)
Comment (elliottprogrammer) [reply client/my-sites/checkout/checkout-thank-you/pending/index.tsx]: Yes, totally, good idea.  Thanks!

File: client/my-sites/checkout/src/lib/pending-page.ts (hunk 3 above)
Comment (elliottprogrammer) [reply client/my-sites/checkout/src/lib/pending-page.ts]: Hmm, yes maybe so..  I mean, it has been working when the `fromSiteSlug` was only in the `if ( receiptId )` block and not the others, but it does make sense that it should probably be added to the `if ( transaction )` block also. So I'm adding it. 👍
All I know is that `fromSiteSlug` should be available here to complete the transaction during the siteless "connect after checkout" flow (`/checkout/jetpack/:product_slug`) when the following 3 url query args are passed from the site -> connectAfterCheckout(true), fromSiteSlug, and adminUrl.

Basically, 'fromSiteSlug' is the "site-slug passed from the site" via url query arg when we do not has the site in context (ie- the site is not connected). 
 As opposed to `siteSlug` which can be retrieved only when the site is in context (which we don't have in the siteless checkout flow).

File: client/my-sites/checkout/src/hooks/use-create-payment-complete-callback.tsx (hunk 8 above)
Comment (elliottprogrammer) [reply client/my-sites/checkout/src/hooks/use-create-payment-complete-callback.tsx:68]: No, `adminUrl: wpAdminUrl` is the `adminUrl` prop passed in, but renamed to `wpAdminUrl` for the component to use.

File: client/my-sites/checkout/src/hooks/use-create-payment-complete-callback.tsx (hunk 9 above)
Comment (elliottprogrammer) [reply client/my-sites/checkout/src/hooks/use-create-payment-complete-callback.tsx:98]: No, `const adminUrl` here is basically equal to: admin_url pulled from state, or if that is undefined/null, then equal to wpAdminUrl passed into the component from props, or if that is also undefined then adminUrl will be undefined.

I supplied more explanation [in the comment below](https://github.com/Automattic/wp-calypso/pull/82036#discussion_r1337605336).

File: client/my-sites/checkout/src/lib/pending-page.ts (hunk 3 above)
Comment (sirbrillig) [reply client/my-sites/checkout/src/lib/pending-page.ts]: I suggest testing the flow with PayPal as the payment method if you haven't already. That flow is different enough from the others (it requires like 4 redirects altogether) that it can highlight bugs where query parameters or state variables might be lost.

File: client/my-sites/checkout/checkout-thank-you/pending/index.tsx [hunk 11]
```diff
@@ -37,6 +37,11 @@ interface CheckoutPendingProps {
 	receiptId: number | undefined;
 	siteSlug?: string;
 	redirectTo?: string;
+	// `fromSiteSlug` is the Jetpack site slug passed from the site via url query arg (into
+	// checkout), for use cases when there is not a site in context, such as siteless checkout.
+	// As opposed to `siteSlug` which is the site slug present when the site is in context
+	// (ie- the site is available in state, such as when site is connected and user logged in).
+	fromSiteSlug?: string;
```
Comment (sirbrillig) [inline client/my-sites/checkout/checkout-thank-you/pending/index.tsx]: This is relatively minor, but at least in my IDE you have to write property comments using multi-line comment style in order to get them to show up when examining the property.

Right now I see this:
Screenshot 2023-09-27 at 11 04 00 AM

But if I apply the following change, I see this instead:
Screenshot 2023-09-27 at 11 02 18 AM




```suggestion
	/**
	 * `fromSiteSlug` is the Jetpack site slug passed from the site via url
	 * query arg (into checkout), for use cases when there is not a site in
	 * context, such as siteless checkout. As opposed to `siteSlug` which is
	 * the site slug present when the site is in context (ie- the site is
	 * available in state, such as when site is connected and user logged in).
	 */
	fromSiteSlug?: string;
```

Comment (elliottprogrammer) [reply client/my-sites/checkout/checkout-thank-you/pending/index.tsx]: Ahh yes, I see. **TIL!**  Ok yes, will do, thanks @sirbrillig!

File: client/my-sites/checkout/src/lib/pending-page.ts (hunk 3 above)
Comment (elliottprogrammer) [reply client/my-sites/checkout/src/lib/pending-page.ts]: I updated the PR so that this "connect after checkout" flow works correctly with PayPal and generic payment processors. 👍

File: client/my-sites/checkout/checkout-thank-you/pending/index.tsx (hunk 1 above)
Comment (elliottprogrammer) [reply client/my-sites/checkout/checkout-thank-you/pending/index.tsx]: ✅

**Review by IanRamosC — APPROVED** [review]
Tests well.
